### PR TITLE
edited-search-sort

### PR DESCRIPTION
**Pull request by younessoudouh**
-edited search by(name, id, desc)
-edited sort by name

### CODE REVIEW

File: src/components/Main.jsx [hunk 1]
```diff
@@ -32,14 +32,14 @@ useEffect(() => {
...
+useEffect(() => {
+  setCustomersToRender(searchedCustomers);
+}, [searchedCustomers])
 
   return (
     <div className="container">
-      <Header props={[customersData, setCustomersToRender, SetCurrentPage]}/>
-      <Table props = {[customersData, customersReadyToRender, setCustomersToRender, setCustomersData]} />
+      <Header props={[sortedCustomers, setSearchedCustomers, SetCurrentPage]}/>
+      <Table props = {[customersData, customersReadyToRender,setSortedCustomers, setCustomersData]} />
```
Comment (faouzioudouh) [inline src/components/Main.jsx]: Why are you passing props like that?
You shouldn't

File: src/components/Search/Search.jsx [hunk 2]
```diff
@@ -22,8 +22,8 @@ const handleChange = (e) => {
 }
 
 useEffect(() => {
-    setCustomersToRender(searchCustomers(customersData));
-},[searchValue])
+    setSearchedCustomers(searchCustomers(sortedCustomers));
+},[searchValue,sortedCustomers])
```
Comment (faouzioudouh) [inline src/components/Search/Search.jsx]: Why do you need `useEffect` here?

File: src/components/Search/Search.jsx [hunk 3]
```diff
@@ -3,14 +3,14 @@ import "./Search.css";
 
 const Search = ({props}) => {
```
Comment (faouzioudouh) [inline src/components/Search/Search.jsx:9]: can you remove this empty line?

File: src/components/Search/Search.jsx [hunk 4]
```diff
@@ -3,14 +3,14 @@ import "./Search.css";
...
 
-let [customersData, setCustomersToRender, SetCurrentPage] = props;
+let [sortedCustomers, setSearchedCustomers, SetCurrentPage] = props;
 
 const [searchValue,setSearchValue] = useState("");
```
Comment (faouzioudouh) [inline src/components/Search/Search.jsx]:
```suggestion
const [searchValue, setSearchValue] = useState("");
```

File: src/components/Search/Search.jsx [hunk 5]
```diff
@@ -3,14 +3,14 @@ import "./Search.css";
...
 
 const searchCustomers = customersToSearchIn => {
     let valueToSearch = searchValue.toLowerCase();
     let searchedCustomers = customersToSearchIn.filter(customer => {
-        return (customer.firstName.toLowerCase()).includes(valueToSearch) || customer.lastName.toLowerCase().includes(valueToSearch) || customer.description.toLowerCase().includes(valueToSearch);
+        return (customer.firstName.toLowerCase().includes(valueToSearch) || customer.lastName.toLowerCase().includes(valueToSearch) || customer.description.toLowerCase().includes(valueToSearch) || customer.id.toString().includes(valueToSearch)) ;
```
Comment (faouzioudouh) [inline src/components/Search/Search.jsx]: Better format please, line too long

File: src/components/Table/Table.jsx [hunk 6]
```diff
@@ -52,8 +52,8 @@ const sortCustomersByName = (customers, sortOrder) => {
 }
 
 useEffect(() => {
-    setCustomersToRender(sortCustomersByName(customersData,sortByName));
-}, [sortByName])
+    setSortedCustomers(sortCustomersByName(customersData,sortByName));
+}, [sortByName,customersData])
```
Comment (faouzioudouh) [inline src/components/Table/Table.jsx]: This is a good usage of `useEffect`.

Sort only when `customersData` or `sortByName` changes

File: src/components/Main.jsx [hunk 7]
```diff
@@ -20,10 +20,10 @@ const getCustomersFromLocalStorage = () => {
 };
 
 const [customersData,setCustomersData] = useState(() => getCustomersFromLocalStorage());
```
Comment (faouzioudouh) [inline src/components/Main.jsx]: Why did you use `useState` initial function, I'm curious

**Review by faouzioudouh — CHANGES_REQUESTED** [review]

File: src/components/Main.jsx (hunk 1 above)
Comment (younessoudouh) [reply src/components/Main.jsx]: I should pass one buy one?

File: src/components/Search/Search.jsx (hunk 3 above)
Comment (younessoudouh) [reply src/components/Search/Search.jsx:9]: Yes sure.

File: src/components/Search/Search.jsx (hunk 5 above)
Comment (younessoudouh) [reply src/components/Search/Search.jsx]: Ok

File: src/components/Search/Search.jsx (hunk 2 above)
Comment (younessoudouh) [reply src/components/Search/Search.jsx]: To change searched customers when search value change or sortedcustomers did

File: src/components/Main.jsx (hunk 7 above)
Comment (younessoudouh) [reply src/components/Main.jsx]: Just to optimize performance I use lazy initialization to get data from local storage only in first rendering

File: src/components/Table/Table.jsx (hunk 6 above)
Comment (younessoudouh) [reply src/components/Table/Table.jsx]: Ok

File: src/components/Main.jsx (hunk 1 above)
Comment (faouzioudouh) [reply src/components/Main.jsx]: Yes please.

File: src/components/Main.jsx (hunk 7 above)
Comment (faouzioudouh) [reply src/components/Main.jsx]: Good

File: src/components/Footer/Footer.jsx [hunk 8]
```diff
@@ -1,52 +1,60 @@
-import React,{useEffect, useState} from "react";
+import React, { useEffect, useState } from "react";
 import "./Footer.css";
 
-const Footer = ({props}) => {
+const Footer = ({ currentPage, rowsPerPage, setRowsPerPage, SetCurrentPage, customersToRender, customersReadyToRender }) => {
```
Comment (faouzioudouh) [inline src/components/Footer/Footer.jsx]: Whats the difference between `customersToRender` and `customersReadyToRender`?

File: src/components/Footer/Footer.jsx [hunk 9]
```diff
@@ -1,52 +1,60 @@
...
 
-let  [currentPage,rowsPerPage,setRowsPerPage,SetCurrentPage,customersToRender,customersReadyToRender] = props;
+    const [activeCustomers, setActiveCustomers] = useState(countActiveCustomers(customersToRender));
 
-const countActiveCustomers = (customersList) => {
-    return customersList.filter(customer=>customer.status==="active").length;
-}  
+    const handleChange = (e) => {
+        const rows = parseInt(e.target.value);
+        SetCurrentPage(1);
```
Comment (faouzioudouh) [inline src/components/Footer/Footer.jsx]:
```suggestion
        setCurrentPage(1);
```

File: src/components/Footer/Footer.jsx [hunk 10]
```diff
@@ -1,52 +1,60 @@
...
                     <li className="rows-page">
                         Rows per page:
                         <select onChange={handleChange}>
                             <option value="5">5</option>
-                            <option value="10">10</option>
+                            <option value="10" selected>10</option>
                             <option value="15">15</option>
                             <option value="20">20</option>
                             <option value="25">25</option>
                         </select>
```
Comment (faouzioudouh) [inline src/components/Footer/Footer.jsx]: I would recommend you to create a component for select.
We might need it somewhere else

File: src/components/Header/Header.jsx [hunk 11]
```diff
@@ -4,17 +4,21 @@ import Button from "../Button/Button";
...
+const Header = ({ sortedCustomers, setSearchedCustomers, SetCurrentPage }) => {
+    return (
         <header>
             <div className="filter-wrap">
                 <Filter />
-                <Search props={props} />
+                <Search
+                    sortedCustomers={sortedCustomers}
+                    setSearchedCustomers={setSearchedCustomers}
+                    SetCurrentPage={SetCurrentPage}
```
Comment (faouzioudouh) [inline src/components/Header/Header.jsx]:
```suggestion
                    setCurrentPage={setCurrentPage}
```

File: src/components/Filter/Filter.jsx [hunk 12]
```diff
@@ -1,16 +1,21 @@
-import React,{useState} from "react";
-import FilterModule from "../FilterModule/FilterModule"
+import React, { useState } from "react";
+import FilterPopUp from "../FilterPopUp/FilterPopUp"
 import "./Filter.css";
 
-const Filter =()=>{
-    const [clicked,setClicked]=useState(false);
+const Filter = ({ sort, setSort }) => {
+    const [filterClicked, setFilterClicked] = useState(false);
```
Comment (faouzioudouh) [inline src/components/Filter/Filter.jsx]:
```suggestion
    const [filterOpen, setFilterOpen] = useState(false);
```

File: src/components/Filter/Filter.jsx [hunk 13]
```diff
@@ -1,16 +1,21 @@
...
 
-    const handleClick =()=> setClicked(!clicked);
+    const handleClick = (e) => {
```
Comment (faouzioudouh) [inline src/components/Filter/Filter.jsx]:
```suggestion
    const handleClick = (event) => {
```

File: src/components/Filter/Filter.jsx [hunk 14]
```diff
@@ -1,16 +1,21 @@
...
+        e.stopPropagation();
+        setFilterClicked(!filterClicked)
```
Comment (faouzioudouh) [inline src/components/Filter/Filter.jsx]:
```suggestion
        setFilterOpen((previousState) => !previousState);
```

File: src/components/Filter/Filter.jsx [hunk 15]
```diff
@@ -1,16 +1,21 @@
...
+    };
 
-    return(
+    window.addEventListener("click", () => setFilterClicked(false));
```
Comment (faouzioudouh) [inline src/components/Filter/Filter.jsx]: `useEffect`

Comment (faouzioudouh) [reply src/components/Filter/Filter.jsx]: Cleanup

Comment (faouzioudouh) [reply src/components/Filter/Filter.jsx]: Whats out for the same function reference

File: src/components/FilterPopUp/FilterPopUp.css [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+    background-color: #fff;
+    z-index: 500;
+    padding: 10px;
+    width: 150px;
+    box-shadow: 2px 2px 10px 4px rgba(46, 46, 51, 0.2);
+    border-radius: 5px;
+    pointer-events: visible;
+}
+
+.sort h3 {
```
Comment (faouzioudouh) [inline src/components/FilterPopUp/FilterPopUp.css]: Don't style tags

File: src/components/Footer/Footer.jsx [hunk 17]
```diff
@@ -1,52 +1,61 @@
-import React,{useEffect, useState} from "react";
+import React, { useEffect, useState } from "react";
 import "./Footer.css";
+import Select from "../Select/Select";
 
-const Footer = ({props}) => {
+const Footer = ({ currentPage, rowsPerPage, setRowsPerPage, setCurrentPage, customersToRender, customersReadyToRender }) => {
```
Comment (faouzioudouh) [inline src/components/Footer/Footer.jsx]:
```suggestion
const Footer = ({ currentPage, rowsPerPage, setRowsPerPage, setCurrentPage, customers, allCustomersCount }) => {
```

File: src/components/Footer/Footer.jsx [hunk 18]
```diff
@@ -1,52 +1,61 @@
...
+    const countActiveCustomers = (customersList) => {
+        return customersList.filter(customer => customer.status === "active").length;
+    }
 
-let  [currentPage,rowsPerPage,setRowsPerPage,SetCurrentPage,customersToRender,customersReadyToRender] = props;
+    const [activeCustomers, setActiveCustomers] = useState(countActiveCustomers(customersToRender));
```
Comment (faouzioudouh) [inline src/components/Footer/Footer.jsx]: remove this useless state

File: src/components/Footer/Footer.jsx [hunk 19]
```diff
@@ -1,52 +1,61 @@
...
+    const [slectedValue, setSlectedValue] = useState(10)
```
Comment (faouzioudouh) [inline src/components/Footer/Footer.jsx]: give it a meaningful name

File: src/components/Footer/Footer.jsx [hunk 20]
```diff
@@ -1,52 +1,61 @@
...
 
-const countActiveCustomers = (customersList) => {
-    return customersList.filter(customer=>customer.status==="active").length;
-}  
+    const handleChange = (e) => {
```
Comment (faouzioudouh) [inline src/components/Footer/Footer.jsx]: meaningful name please

File: src/components/Footer/Footer.jsx [hunk 21]
```diff
@@ -1,52 +1,61 @@
...
+        setActiveCustomers(countActiveCustomers(customersToRender));
+    }, [customersToRender])
 
     return (
         <footer>
             <div className="active-customers">
-                active customers: <strong>{activeCustomers}</strong> / <small>{customersToRender.length}</small>
+                active customers:
+                {` `}
+                <strong>{activeCustomers}</strong>
```
Comment (faouzioudouh) [inline src/components/Footer/Footer.jsx]:
```suggestion
                <strong>{activeCustomersCount}</strong>
```

File: src/components/Footer/Footer.jsx [hunk 22]
```diff
@@ -1,52 +1,61 @@
...
-                        <select onChange={handleChange}>
-                            <option value="5">5</option>
-                            <option value="10">10</option>
-                            <option value="15">15</option>
-                            <option value="20">20</option>
-                            <option value="25">25</option>
-                        </select>
+                        <Select
+                            options={[5, 10, 15, 20, 25]}
+                            slectedValue={slectedValue}
```
Comment (faouzioudouh) [inline src/components/Footer/Footer.jsx]:
```suggestion
                            selectedValue={selectedValue}
```

File: src/components/Footer/Footer.jsx (hunk 19 above)
Comment (faouzioudouh) [inline src/components/Footer/Footer.jsx]: Duplicated?

File: src/components/Main.jsx [hunk 23]
```diff
@@ -2,45 +2,67 @@ import Header from './Header/Header';
...
+      setCurrentPage(previous => previous - 1);
+    }
+  }, [customersReadyToRender])
 
-// useEffect(() => {
-//   setCustomersToRender(customersData);
-// }, [customersData])
+  useEffect(() => {
+    setCustomersToRender(searchedCustomers);
+  }, [searchedCustomers])
```
Comment (faouzioudouh) [inline src/components/Main.jsx]: Is it needed?

File: src/components/Search/Search.jsx [hunk 24]
```diff
@@ -1,34 +1,36 @@
...
-    return searchedCustomers;
-}
-
-const handleChange = (e) => {
-    setSearchValue(e.target.value);
-    SetCurrentPage(1);
-}
-
-useEffect(() => {
-    setCustomersToRender(searchCustomers(customersData));
```
Comment (faouzioudouh) [inline src/components/Search/Search.jsx:25]: Do you need it?

File: src/components/Table/Table.jsx [hunk 25]
```diff
@@ -1,81 +1,97 @@
...
+
+        return sortCustomersByStatus(sortedByName, sortOrders);
+    }
 
-useEffect(() => {
-    setCustomersToRender(sortCustomersByName(customersData,sortByName));
-}, [sortByName])
+    useEffect(() => {
+        setSortedCustomers(sortCombined(customersData, sort));
+    }, [sort, customersData])
```
Comment (faouzioudouh) [inline src/components/Table/Table.jsx]: Do you need it?

**Review by faouzioudouh — APPROVED** [review]
Great work!

**Review by faouzioudouh — APPROVED** [review]

**Pull request merged**